compiler/crates/relay-codegen/src/build_ast.rs

```diff
@@ -744,15 +849,41 @@ impl<'schema, 'builder> CodegenBuilder<'schema, 'builder> {
             CODEGEN_CONSTANTS.relay_resolver
         };
 
-        // TODO(T86853359): Support non-haste environments when generating Relay Resolver Reader AST
-        let haste_import_name = Path::new(&module.to_string())
-            .file_stem()
-            .unwrap()
-            .to_string_lossy()
-            .intern();
-
         let args = self.build_arguments(field_arguments);
 
+        let definition_artifact_location = path_for_artifact(
+            self.project_config,
+            match definition {
+                ExecutableDefinition::Operation(operation) => {
+                    operation.name.location.source_location()
+                }
+                ExecutableDefinition::Fragment(fragment) => {
+                    fragment.name.location.source_location()
+                }
+            },
+            match definition {
+                ExecutableDefinition::Operation(operation) => operation.name.item,
+                ExecutableDefinition::Fragment(fragment) => fragment.name.item,
+            },
+        );
+
+        let module_location = PathBuf::from_str(module.lookup()).unwrap();
+
+        let module_path = module_location.parent().unwrap();
+        let definition_artifact_location_path = definition_artifact_location.parent().unwrap();
+
+        println!(
+            "Diffing paths: {:?} : {:?}",
+            definition_artifact_location_path, module_path
+        );
+
+        let resolver_module_location =
+            pathdiff::diff_paths(module_path, definition_artifact_location_path).unwrap();
+
+        let import_path = resolver_module_location.join(module_location.file_name().unwrap());
```

Comment:
Obviously we need to check the config and use the Haste name if that's how the project is configured, but I assume that's outside of the scope of this sketch.